### PR TITLE
blob: use new URLMux pattern for OpenBucket

### PR DESCRIPTION
**Pull request by zombiezen**
Design discussion in #1209.

A follow-up PR will shorten the tutorial to use `blob.OpenBucket`.

Updates #1174

### CODE REVIEW

File: blob/azureblob/azureblob.go [hunk 1]
```diff
@@ -17,33 +17,11 @@
...
-//
-// The URL's Host is used as the bucket name.
-//
-// By default, credentials are retrieved from the environment variables
-// AZURE_STORAGE_ACCOUNT, AZURE_STORAGE_KEY, and AZURE_STORAGE_SAS_TOKEN.
-// AZURE_STORAGE_ACCOUNT is required, along with one of the other two. See
+// For blob.OpenBucket URLs, azureblob obtains credentials from the environment
+// variables AZURE_STORAGE_ACCOUNT, AZURE_STORAGE_KEY, and
+// AZURE_STORAGE_SAS_TOKEN. AZURE_STORAGE_ACCOUNT is required, along with one of
+// the other two. For more on SAS tokens, see
```
Comment (vangent) [inline blob/azureblob/azureblob.go]: Can you add back something to the comment with:

1. What the scheme is.
2. A sample URL.

?

File: blob/azureblob/azureblob_test.go [hunk 2]
```diff
@@ -264,113 +261,3 @@ func TestOpenBucket(t *testing.T) {
 		})
 	}
 }
-
-func TestOpenURL(t *testing.T) {
```
Comment (vangent) [inline blob/azureblob/azureblob_test.go]: Why did you remove this entirely? It looks like test coverage on Travis went way down as a result. I think you can basically put it back, just removing the tests that were focused on `cred_path`.

File: blob/blob.go [hunk 3]
```diff
@@ -833,65 +831,74 @@ type WriterOptions struct {
...
-var (
-	// registry maps scheme strings to provider-specific instantiation functions.
-	registry = map[string]FromURLFunc{}
-	// registryMu protected registry.
-	registryMu sync.Mutex
-)
+// A type that implements BucketURLOpener can open buckets based on a URL.
+// The opener must not modify the URL argument. OpenBucketURL must be safe to
+// call from multiple goroutines.
+type BucketURLOpener interface {
```
Comment (vangent) [inline blob/blob.go]: Add back "It is intended for use by provider implementations."?

I think we should try to make it clear to `blob.Bucket` users which parts of the package they can/should ignore.

File: blob/blob.go [hunk 4]
```diff
@@ -833,65 +831,74 @@ type WriterOptions struct {
...
+func (mux *URLMux) OpenBucketURL(ctx context.Context, u *url.URL) (*Bucket, error) {
 	if u.Scheme == "" {
-		return nil, fmt.Errorf("invalid URL %q, missing scheme", urlstr)
+		return nil, fmt.Errorf("open bucket %q: no scheme in URL", u)
 	}
-	fn := fromRegistry(u.Scheme)
-	if fn == nil {
-		return nil, fmt.Errorf("no provider registered for scheme %q", u.Scheme)
+	var opener BucketURLOpener
+	if mux != nil {
```
Comment (vangent) [inline blob/blob.go]: Nit: is this `nil` check needed? Seems OK to panic here to me.

File: blob/blob.go [hunk 5]
```diff
@@ -833,65 +831,74 @@ type WriterOptions struct {
...
+		return nil, fmt.Errorf("open bucket %q: no provider registered for %s", u, u.Scheme)
 	}
-	return NewBucket(drv), nil
+	return opener.OpenBucketURL(ctx, u)
+}
+
+var defaultURLMux = new(URLMux)
+
+// DefaultURLMux returns the URLMux used by OpenBucket.
+func DefaultURLMux() *URLMux {
```
Comment (vangent) [inline blob/blob.go]: IIUC, this is also "intended for use by provider implementations."

File: blob/fileblob/fileblob.go [hunk 6]
```diff
@@ -30,25 +30,6 @@
 // (e.g., "~", which unescapes to "~", which escapes back to "%7E" != "~"),
 // aren't visible using fileblob.
 //
-// Open URLs
```
Comment (vangent) [inline blob/fileblob/fileblob.go]: Put part of this back? IMHO each driver package should have a `Open URLs` section (or maybe `OpenBucket URLs` now) that at least describe the scheme and a sample URL. It can refer to the `URLOpener` type for more details.

File: blob/gcsblob/gcsblob.go [hunk 7]
```diff
@@ -15,23 +15,6 @@
 // Package gcsblob provides a blob implementation that uses GCS. Use OpenBucket
 // to construct a *blob.Bucket.
 //
-// Open URLs
```
Comment (vangent) [inline blob/gcsblob/gcsblob.go]: Same.

File: blob/gcsblob/gcsblob.go [hunk 8]
```diff
@@ -70,48 +54,85 @@ import (
...
+type URLOpener struct {
+	// Client must be set to a non-nil HTTP client authenticated with
+	// Cloud Storage scope or equivalent.
+	Client *gcp.HTTPClient
+
+	// Options specifies the default options to pass to OpenBucket.
+	Options Options
+}
+
+// OpenBucketURL opens the GCS bucket with the same name of the URL's host.
```
Comment (vangent) [inline blob/gcsblob/gcsblob.go]: Nit: "of" -> "as"?

File: blob/gcsblob/gcsblob.go [hunk 9]
```diff
@@ -70,48 +54,85 @@ import (
...
+func (o *URLOpener) OpenBucketURL(ctx context.Context, u *url.URL) (*blob.Bucket, error) {
+	var err error
+	o, err = o.forParams(ctx, u.Query())
```
Comment (vangent) [inline blob/gcsblob/gcsblob.go]: It took me a while to parse what's going on here (it's kind of weird to overwrite `o`). Could it just be done inline, and only copying/update `Options` instead of a new `URLOpener`?

File: blob/gcsblob/gcsblob.go (hunk 7 above)
Comment (vangent) [reply blob/gcsblob/gcsblob.go]: In particular, the docstring should explain that by default `OpenBucket` will use default GCP credentials, and link to https://cloud.google.com/docs/authentication/production.

File: blob/gcsblob/gcsblob_test.go [hunk 10]
```diff
@@ -291,92 +291,46 @@ func TestOpenURL(t *testing.T) {
...
-	if _, err := credFile.Write(jsonCred); err != nil {
-		t.Fatal(err)
-	}
-	if err := credFile.Close(); err != nil {
-		t.Fatal(err)
-	}
-
-	// Set invalid (and later restore) default credentials in the environment,
-	// so that this test always fails.
-	prevEnv := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")
```
Comment (vangent) [inline blob/gcsblob/gcsblob_test.go]: I think you lost some coverage by removing this...?

File: blob/memblob/memblob.go [hunk 11]
```diff
@@ -15,15 +15,6 @@
 // Package memblob provides an in-memory blob implementation.
 // Use OpenBucket to construct a *blob.Bucket.
 //
-// Open URLs
```
Comment (vangent) [inline blob/memblob/memblob.go]: Same as above.

File: blob/s3blob/s3blob.go [hunk 12]
```diff
@@ -15,22 +15,6 @@
 // Package s3blob provides a blob implementation that uses S3. Use OpenBucket
 // to construct a *blob.Bucket.
 //
-// Open URLs
```
Comment (vangent) [inline blob/s3blob/s3blob.go]: Same as above.

File: blob/s3blob/s3blob.go [hunk 13]
```diff
@@ -71,30 +56,53 @@ import (
...
 
-func openURL(ctx context.Context, u *url.URL) (driver.Bucket, error) {
-	q := u.Query()
-	cfg := &aws.Config{}
+// URLOpener opens S3 URLs like "s3://mybucket".
+type URLOpener struct {
+	// ConfigProvider must be set to a non-nil value.
+	ConfigProvider client.ConfigProvider
 
-	if region := q["region"]; len(region) > 0 {
```
Comment (vangent) [inline blob/s3blob/s3blob.go]: Can you put these back into the default implementation?

File: blob/s3blob/s3blob_test.go [hunk 14]
```diff
@@ -221,45 +220,3 @@ func TestOpenBucket(t *testing.T) {
 		})
 	}
 }
-
-func TestOpenURL(t *testing.T) {
```
Comment (vangent) [inline blob/s3blob/s3blob_test.go]: Why remove?

**Comment by vangent** [comment]
Can you sync to HEAD as well? Looks like there are some conflicts, hopefully not too hairy.

**Comment by zombiezen** [comment]
PTAL

File: go.mod [hunk 15]
```diff
@@ -15,11 +15,15 @@
 module gocloud.dev
 
 require (
-	cloud.google.com/go v0.35.0
+	cloud.google.com/go v0.34.0
```
Comment (vangent) [inline go.mod]: Looks like you need to revert your `go.mod` and `go.sum` changes.

File: blob/fileblob/fileblob.go [hunk 16]
```diff
@@ -15,24 +15,26 @@
 // Package fileblob provides a blob implementation that uses the filesystem.
 // Use OpenBucket to construct a *blob.Bucket.
 //
+// Blob keys are escaped before being used as filenames, and filenames are
```
Comment (vangent) [inline blob/fileblob/fileblob.go]: Looks like you added this couple of paragraphs back; please delete (it's replaced by the `Escaping` section below).

File: go.mod (hunk 15 above)
Comment (zombiezen) [reply go.mod]: Ack, I messed up the Git command to do so. Sigh...

File: blob/fileblob/fileblob.go (hunk 16 above)
Comment (zombiezen) [reply blob/fileblob/fileblob.go]: Done.

File: blob/s3blob/s3blob.go (hunk 13 above)
Comment (vangent) [reply blob/s3blob/s3blob.go]: I failed to follow up on this comment. I think it's important to include these, will send a PR.
